index.js

```diff
@@ -666,11 +666,35 @@ class ServerlessAppsyncPlugin {
 
     // if there are substitutions for this template then add fn:sub
     if (Object.keys(substitutions).length > 0) {
-      return { 'Fn::Sub': [template, substitutions] };
+      return this.substituteGlobalTemplateVariables(template, substitutions);
     }
 
     return template;
   }
+
+  /**
+   * Creates Fn::Join object from given template where all given substitutions
+   * are wrapped in Fn::Sub objects. This enables template to have also 
+   * characters that are not only alphanumeric, underscores, periods, and colons.
+   * 
+   * @param {*} template 
+   * @param {*} substitutions 
+   */
+  substituteGlobalTemplateVariables(template, substitutions) {
+    let keys = Object.keys(substitutions);
+    for (let i = 0; i < keys.length; i++) {
+      template = template.split('${' + keys[i] + '}').join('||' + keys[i] + '||');
```

Comment:
I tested it and `$variableName` does not get replaced now either. For some time I thought that this is a workaround for this bug but it did not replace those variables at all.

Reply:
@deankostomaj please ignore this previous comment.
After understanding the problem, what you did looks good.
However, I would still consider using `replace` instead of the split/join.

```suggestion
      template = template.replace(/\${(.*?)}/g, '||$1||');
```

Also, do you think we should use something else than `||` as separator? This could be confused with the `or` operator, even worse, in case of using the `or` operator in a template (though that I am not sure if this is possible), this could break the template.

Maybe something more "unique" would work better? At worse, we could use `|||` (3 of them) ?

Reply:
Overall the problem was that before substitutions were made with `{'Fn::Sub': [template, substitutions]}` and because the whole template was in variable `template` AWS threw an exception that only `alphanumeric characters, underscores, periods, and colons` can be in `Fn::Sub`. I solved this problem by parsing template and using `Fn::Join` and joined pieces of template with `Fn::Sub` where I passed in only variable and its value. This was no special characters error will be thrown.

Reply:
@deankostomaj Yes, `$variableName` should be kept as is and never replaced as they are interpreted by VTL. 
So all good.

Reply:
@deankostomaj yes, this is what I figured out. All is OK :D

Reply:
OK. For some reason github did not update comments and I did not see your comments until now. Yes maybe we could use something else than `||`. Maybe we could use your suggestion `|||`.

Reply:
> @deankostomaj please ignore this previous comment.
> After understanding the problem, what you did looks good.
> However, I would still consider using `replace` instead of the split/join.
> 
> Suggested change Beta 
> [Give us feedback](https://www.research.net/r/SuggestedChanges)
>       template = template.split('${' + keys[i] + '}').join('||' + keys[i] + '||');
>       template = template.replace(/\${(.*?)}/g, '||$1||');
> Commit suggestion  Add suggestion to batch
> Also, do you think we should use something else than `||` as separator? This could be confused with the `or` operator, even worse, in case of using the `or` operator in a template (though that I am not sure if this is possible), this could break the template.
> 
> Maybe something more "unique" would work better? At worse, we could use `|||` (3 of them) ?

@bboure would `replace` replaced all occurrences?

Reply:
@deankostomaj 

> would `replace` replaced all occurrences?

With `.*`, yes. And this is NOT what we want here. We need to use something like 

````javascript
variables = keys.join("|");
regex = new RegExp("\\${(" + variables + ")}", "g")
template = template.replace(regex, "|||$1|||");
````

Reply:
OK. I get it. It would be more elegant solution. I will fix that.

Reply:
@bboure I included your suggestion.

Reply:
@deankostomaj Looks good. 
Can variables contain some special characters that could break the regex?
Like dots for example?

Reply:
@bboure I'm not sure. As far as I know variables are all letter.

Reply:
@deankostomaj yeah, I guess it makes sense.

Reply:
@deankostomaj Did you test this? Are you confident? can I merge? 😄

Reply:
Yes. I test it with multiple variablres, but I did not test it with multiple occurrences of one variable. I'll do also that and I'll let you know.

Reply:
@bboure I have done some more testing and all seems OK. I have tested with multiple variables and multiple occurrences.